mne_bids/datasets.py

```diff
@@ -5,15 +5,156 @@
 #          Stefan Appelhoff <stefan.appelhoff@mailbox.org>
 #
 # License: BSD (3-clause)
-
 import os
 import os.path as op
 import shutil
 import tarfile
-import urllib.request
+import urllib
 
 from mne.utils import _fetch_file
 
+from mne_bids import make_bids_basename, make_bids_folders
+
+
+def _download_data(data, overwrite, verbose, dry=False):
+    """Iterate over `data`, a dict with fpath, url and download."""
+    for fpath, url in data.items():
+        if dry:
+            print(fpath, url)
+            continue
+        if op.exists(fpath) and not overwrite:
+            continue
+        _fetch_file(url=url, file_name=fpath, print_destination=verbose,
+                    resume=True, timeout=10., verbose=verbose)
+
+
+def fetch_matchingpennies(data_path=None, subjects=None, dataset_data=True,
+                          sourcedata=False, overwrite=False, verbose=True,
+                          dry=False):
```

Comment:
With a little bit of work, this could be made into a nice fetcher for BIDS datasets. Something like:

```py
def fetch_bids_dataset(data_path, url, subject, run, task):
    pass
```

Reply:
The issue is that the URL "Keys" were extracted manually from OSF, because the automatic scrapers were failing.

So it cannot be a generic fetcher without storing the "key" data somewhere.

in MNE-Python this is "circumvented" by zipping datasets, so they have a single URL ... of course, this comes at the expense that you cannot download single subjects.

Reply:
How does MNE download the EEGBCI datasets? See: https://martinos.org/mne/stable/generated/mne.datasets.eegbci.load_data.html? OSF authentication is a different issue. We won't host BIDS datasets there. Do you have the same issue with Openneuro? For eeg_matching_pennies, you could upload it to openneuro ...

Reply:
> How does MNE download the EEGBCI datasets? 

The EEGBCI dataset is hosted on a different website, which supports sensible URLs that can be "guessed" (see [here](https://github.com/mne-tools/mne-python/blob/34bd8b605b459da3cbb2b2657862ed09c16fdfcb/mne/datasets/eegbci/eegbci.py#L162-L163) ... on OSF, there is no file-tree apparent in the URL, instead - each file seems to have some kind of hash that might be random. 

> OSF authentication is a different issue.

authentification is no issue, because we are not downloading private data (only public)

> Do you have the same issue with Openneuro? 

No, for openneuro, we can use datalad

> For eeg_matching_pennies, you could upload it to openneuro ...

For eeg_matchingpennies we now have github.com/sappelhoff/eeg_matchingpennies :-) no need for openneuro

Reply:
right, so I see two paths:

1. Use datalad to fetch the datasets. But the problem is dependency and we fight dependencies hard. You've already noticed lots of issues that are broken with datalad. It's fine using it in the CIs but exposing it to the user means we have to make a judgement whether we are okay having it as a dependency.

2. Make a generic fetcher for openneuro in MNE. I think this could be really useful for many people in the MNE community to experience BIDS without having to figure out datalad.

Reply:
> Use datalad to fetch the datasets. But the problem is dependency and we fight dependencies hard.

yes, I also don't like dependencies :)

> Make a generic fetcher for openneuro in MNE. I think this could be really useful for many people in the MNE community to experience BIDS without having to figure out datalad.

Yes, but how do you imagine this to work? Fetching one full dataset at a time? With openneuro, there is also no way to get single subject data ... one could use the Amazon AWS SDK and the [python binding](https://github.com/boto/boto3) ... to access and download full datasets, see [this post](https://openneuro.freshdesk.com/support/solutions/articles/29000023186-issues-with-downloading-datasets) - but as I said, it only downloads the full data. And it would also be an additional dependency

To get single subject data from openneuro, I am afraid that one *has* to go the datalad route and rely on git-annex repos that are automatically created from openneuro

The only "alternative" I see to that would be to upload single subject datasets to OpenNeuro, ... like the MNE-sample-data. But these have their limitations again (because they are small and only serve for specific examples)

Reply:
okay this is a good point.

> With openneuro, there is also no way to get single subject data ... one could use the Amazon AWS SDK and the python binding ... to access and download full datasets, see this post - but as I said, it only downloads the full data.

Based on this post, I understood the opposite -- that you _can_ download parts of the dataset. Did I read it incorrect?

> And it would also be an additional dependency

But it's better than datalad because datalad depends on the AWS CLI: https://github.com/datalad/datalad/blob/master/setup.py#L68. The AWS CLI itself is pretty lightweight but datalad has a crazy number of dependencies ...

Reply:
> Based on this post, I understood the opposite -- that you can download parts of the dataset. Did I read it incorrect?

I am pretty sure that you cannot download individual data, but I'll clarify with Poldracklab! :)

Reply:
Try:

```sh
$ aws s3 sync --no-sign-request s3://openneuro.org/ds000248 ds000248 --exclude "*" --include "*.json"
```

Reply:
okay, thanks for reminding me that we are practical people ... try it before overthinking it. :-D

This is awesome!